code/service/src/main/scala/com/namely/chiefofstate/ServiceMigrationRunner.scala

```diff
@@ -0,0 +1,125 @@
+ *   <ol>
+ *     <li> check the existence of the cos_migrations table
+ *     <li> if the table exists go to step 4
+ *     <li> if the table does not exist run the whole migration and reply back to the [[ServiceBootstrapper]]
+ *     <li> check the current version against the available versions
+ *     <li> if current version is the last version then no need to run the migration just reply back to [[ServiceBootstrapper]]
+ *     <li> if not then run the whole migration and reply back to the [[ServiceBootstrapper]]
+ *   </ol>
+ * </p>
+ */
+object ServiceMigrationRunner {
+  final val log: Logger = LoggerFactory.getLogger(getClass)
+
+  def apply(config: Config): Behavior[ScalaMessage] = Behaviors
+    .setup[ScalaMessage] { context =>
+      Behaviors.receiveMessage[ScalaMessage] {
+        case MessageWithActorRef(message, replyTo) =>
+          message match {
+            case _: DoMigration =>
+              val result: Try[Unit] = {
+                // create and run the migrator
+                val journalJdbcConfig: DatabaseConfig[JdbcProfile] =
+                  JdbcConfig.journalConfig(config)
+
+                // for the migration, as COS projections has moved to raw JDBC
+                // connections and no longer uses slick.
+                val projectionJdbcConfig: DatabaseConfig[JdbcProfile] =
+                  JdbcConfig.projectionConfig(config)
+
+                // get the projection config
+                val priorProjectionJdbcConfig: DatabaseConfig[JdbcProfile] =
+                  JdbcConfig.projectionConfig(config, "chiefofstate.migration.v1.slick")
+
+                // get the old table name
+                val priorOffsetStoreName: String =
+                  config.getString("chiefofstate.migration.v1.slick.offset-store.table")
+
+                val v1: V1 = V1(journalJdbcConfig, priorProjectionJdbcConfig, priorOffsetStoreName)
+                val v2: V2 = V2(journalJdbcConfig, projectionJdbcConfig)(context.system)
+                val v3: V3 = V3(journalJdbcConfig)
+
+                // instance of the migrator
+                val migrator: Migrator = new Migrator(journalJdbcConfig)
+                  .addVersion(v1)
+                  .addVersion(v2)
+                  .addVersion(v3)
+
+                // check whether the migration table exists or not
+                if (!DbUtil.tableExists(journalJdbcConfig, Migrator.COS_MIGRATIONS_TABLE)) {
+                  migrator.run()
+                } else {
+                  // get the last version
+                  val lastVersion: Option[Int] = migrator.getVersions().lastOption.map(_.versionNumber)
+
+                  // get the current version
+                  val currentVersion: Option[Int] = Migrator.getCurrentVersionNumber(journalJdbcConfig)
+
+                  // get the ordering
+                  val ordering: Ordering[Option[Int]] = implicitly[Ordering[Option[Int]]]
+
+                  if (ordering.lt(currentVersion, lastVersion)) {
```

Comment:
This is an example of something only the migrator should understand

Reply:
TODO to be tackled in another PR